### DISCUSSION TITLE
Serverless image service

### DISCUSSION THREAD

**Discussion by yamiteru** [JonasKruckenberg/imagetools#26] [Ideas]
I know, I know! This library is for ahead of time image compression. But please hear me out 🙏

I think that if the `wasm-vips` is integrated into the `core` we might create adapters for different serverless providers such as Cloudflare Workers so anyone can set up an on demand image compression if ahead of time compression is not enough for them.

This library is extremely useful for small to medium SSG websites. But is useless for big SSR e-commerces and platforms. It'd be awesome if this library and its directives could be used freely in any environment.

**Comment by JonasKruckenberg** [comment]
Yes I thought about this too, that's actually one of the reasons for #17. I'm unsure about a few things though:

1. Performance

Sure javascript is not as slow as it was and nodejs has improved quite a lot, but transforming images is quite slow.
While this is fine for SSG, making it work in real-time does require some optimization work.

3. Serverless compatibility

Maybe not as important, because directives are just pure functions and the core library would be treeshakeable but I serverless has a lot of different requirements than traditional servers so something to keep in mind.

2. Security

I've discovered [imgproxy](https://github.com/imgproxy/imgproxy) while researching for this and they're quite concerned with security since transforming images with unknown contents opens you up to a whole bunch of security issues, a lot of those would be a least somewhat adressed by using wasm, but it still something to keep in mind.

So it's not a all a no, it's a great idea but there are things to consider 👍

**Reply by yamiteru** [reply]
Yeah I thought so.

1. Performance
Yes it is slow. Wasm should help a bit with this.

2. Serverless compatibility
I've tried several serverless providers and the only "difficult" and almost always specific to the provider is routing. If we let the routing part up to the dev it should be very easy. Image transformation usually takes quite some cpu time so most of the serverless providers are not a good fit for this. Cloudflare Workers Unbound (https://blog.cloudflare.com/introducing-workers-unbound/) seems to be the best (in term of routing) and least expensive serverless option for image transformation.

If we use it in Node.js on a custom VPS there shouldn't be any problem and we can make it faster by just adding more servers which cannot be done with serverless. Personally I'd choose uWebSockets since it's the fastest Node.js routing library.

3. Security
I don't think this library should be concerned about security. Security is a part of routing and as I said I think it's up to the dev not this library. But we might create starters (Node.js, Cloudflare Workers, AWS Lambda, etc.) with security in mind.

I'm really hyped to see the #17 and #15 in production. I'll be more than happy to help with the Node.js and Cloudflare implementations and starters.
